gixy/plugins/regex_redos.py

```diff
@@ -50,20 +50,58 @@ def __init__(self, config):
         super(regex_redos, self).__init__(config)
         self.redos_server = self.config.get('url')
 
+        self.cached_results = {}
+
     def audit(self, directive):
         # If we have no ReDoS check URL, skip.
         if not self.redos_server:
             return
 
-        # Only process directives that have regex modifiers.
-        if directive.modifier not in ('~', '~*'):
+        regex_patterns = set()
+        child_val = {}
+
+        if directive.name == 'server_name':
+            for server_name in directive.args:
+                if server_name[0] != '~':
+                    continue
+                # server_name www.example.com ~^www.\d+\.example\.com$;
+                regex_patterns.add(server_name[1:])
+        elif directive.name == 'proxy_redirect':
+            if directive.args[0][0] == '~':
+                # proxy_redirect ~^//(.*)$ $scheme://$1;
+                if directive.args[0][1] == '*':
+                    regex_patterns.add(directive.args[0][2:])
+                else:
+                    regex_patterns.add(directive.args[0][1:])
+        elif directive.name == 'location':
+            if directive.modifier in ['~', '~*']:
+                # location ~ ^/example$ { }
+                regex_patterns.add(directive.path)
+        elif directive.name == 'if':
+            if directive.operand in ['~', '~*', '!~', '!~*']:
+                # if ($http_referer ~* ^https://example\.com/$") { }
+                regex_patterns.add(directive.value)
+        elif directive.name == 'rewrite':
+            # rewrite ^/1(/.*) $1 break;
+            regex_patterns.add(directive.pattern)
+        elif isinstance(directive, MapBlock):
```

Comment:
This incorrectly checks every type of map{}, regardless of whether it's `location', 'server_name', 'proxy_redirect', 'if', 'rewrite'. This is kind of deliberate, but I wonder if it can be handled better, i.e. if it could be possible to resolve the variable, determine that it resolves to a Mapblock, and then loop over the Mapblock's children..